micro_dl/light/data.py

```diff
@@ -0,0 +1,215 @@
+
+import numpy as np
+import torch
+from iohub.ngff import ImageArray, Plate, open_ome_zarr
+from lightning.pytorch import LightningDataModule
+from monai.transforms import (
+    CenterSpatialCropd,
+    Compose,
+    NormalizeIntensity,
+    RandAdjustContrastd,
+    RandAffined,
+    RandGaussianSmoothd,
+    RandWeightedCropd,
+)
+from torch.utils.data import DataLoader, Dataset, Subset
+
+
+class SlidingWindowDataset(Dataset):
+    def __init__(
+        self,
+        plate: Plate,
+        source_channel: str,
+        target_channel: str,
+        z_window_size: int,
+        transform: Callable = None,
+        target_center_slice_only: bool = True,
+        normalize_intensity: bool = True,
+    ) -> None:
+        super().__init__()
+        self.plate = plate
+        self.source_ch_idx = plate.get_channel_index(source_channel)
+        self.target_ch_idx = plate.get_channel_index(target_channel)
+        self.z_window_size = z_window_size
+        self.transform = transform
+        self.target_center_slice_only = target_center_slice_only
+        self.normalize_intensity = normalize_intensity
+        self._count_windows()
+        self._get_normalizer(source_channel, target_channel)
+
+    def _count_windows(self) -> None:
+        w = 0
+        self.windows = {}
+        for _, fov in self.plate.positions():
+            ts = fov["0"].frames
+            zs = fov["0"].slices - self.z_window_size + 1
+            w += ts * zs
+            self.windows[w] = fov
+        self._max_window = w
+
+    def _get_normalizer(self, source_channel: str, target_channel: str):
+        norm_meta = self.plate.zattrs["normalization"]
+        self.source_normalizer = NormalizeIntensity(
+            subtrahend=norm_meta[source_channel]["dataset_statistics"]["median"],
+            divisor=norm_meta[source_channel]["dataset_statistics"]["iqr"],
+        )
+        self.target_normalizer = NormalizeIntensity(
+            subtrahend=norm_meta[target_channel]["dataset_statistics"]["median"],
+            divisor=norm_meta[target_channel]["dataset_statistics"]["iqr"],
+        )
+
+    def _find_window(self, index: int) -> tuple[int, int]:
+        window_keys = list(self.windows.keys())
+        window_idx = sorted(window_keys + [index + 1]).index(index + 1)
+        w = window_keys[window_idx]
+        tz = index - window_keys[window_idx - 1] if window_idx > 0 else index
+        return w, tz
+
+    def _read_img_window(self, img: ImageArray, ch_idx: int, tz: int) -> torch.Tensor:
+        zs = img.slices - self.z_window_size + 1
+        t = (tz + zs) // zs - 1
+        z = tz - t * zs
+        selection = (int(t), int(ch_idx), slice(z, z + self.z_window_size))
+        data = img[selection][np.newaxis]
+        return torch.from_numpy(data)
+
+    def __len__(self) -> int:
+        return self._max_window
+
+    def __getitem__(self, index: int) -> dict[str, torch.Tensor]:
+        w, tz = self._find_window(index)
```

Comment:
If I'm understanding this correctly this means we would sample one "valid" patch from each position each epoch (assuming batch size of 1)?

Reply:
Yes, and it's also true when the batch size is larger than 1.